File: Editorial/2_Add_Two_Numbers/TypeScript/1_Iterative_Dummy_Head.ts
function addTwoNumbers(
    l1: ListNode | null,
    l2: ListNode | null
): ListNode | null {
    const dummy = new ListNode(0);
    let current = dummy;
    let carry = 0;

    while (l1 !== null || l2 !== null || carry !== 0) {
        let total = carry;

        if (l1 !== null) {
            total += l1.val;
            l1 = l1.next;
        }

        if (l2 !== null) {
            total += l2.val;
            l2 = l2.next;
        }

        carry = Math.floor(total / 10);
        current.next = new ListNode(total % 10);
        current = current.next;
    }

    return dummy.next;
}

File: Editorial/2_Add_Two_Numbers/CPP/1_Iterative_Dummy_Head.cpp
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode* current = &dummy;
        int carry = 0;

        while (l1 || l2 || carry) {
            int total = carry;

            if (l1) {
                total += l1->val;
                l1 = l1->next;
            }

            if (l2) {
                total += l2->val;
                l2 = l2->next;
            }

            carry = total / 10;
            current->next = new ListNode(total % 10);
            current = current->next;
        }

        return dummy.next;
    }
};
